Main.hs

```diff
@@ -3,27 +3,87 @@ module Main where
 import Control.Monad (unless)
 import Text.Printf (printf)
 
-data Expr
+data Operator = Add 
+     | Sub 
+     | Mult 
+     | Div 
+     | Pow
+     deriving (Eq)
+     
+instance Show Operator where
+     show Add = "+"
+     show Sub = "-"
+     show Mult = "*"
+     show Div = "/"
+     show Pow = "^"
 
-instance Show Expr where
-  show = undefined
+data Expr = Num Double 
+     | Sqrt Expr 
+     | CompExpr Operator Expr Expr
 
+instance Show Expr where
+    show (Num a) = show a
+    show (Sqrt a) = printf "sqrt(%s)" (show a)
+    show (CompExpr op a b) = printf "(%s %s %s)" (show a) (show op) (show b)
+    
 instance Eq Expr where
-  (==) = undefined
+  Num a == Num b = a == b
+  Sqrt a == Sqrt b = a == b
+  CompExpr op a b == CompExpr op' a' b' = op == op' && a == a' && b == b
+  _ == _ = False
 
-data Error
+data Error = NegativeSqrt
+     | ZeroDiv
 
 instance Show Error where
-  show = undefined
+  show NegativeSqrt = printf "Error: Square root of a negative number"
+  show ZeroDiv = printf "Error: Division by 0 is not possible"
 
 instance Eq Error where
-  (==) = undefined
+  NegativeSqrt == NegativeSqrt = True
+  ZeroDiv == ZeroDiv = True
+  _ == _ = False
 
 eval :: Expr -> Either Error Double
-eval = undefined
+eval (Num x) = Right x
+eval (Sqrt expr) =
+   case eval expr of
+      Right x
+         | x >= 0 -> Right (sqrt x)
+         | otherwise -> Left NegativeSqrt
+      Left err -> Left err
+eval (CompExpr op a b) = 
+   case (eval a, eval b) of
+      (Right val1, Right val2) -> 
+         case op of
+            Add  -> Right (val1 + val2)
+            Sub  -> Right (val1 - val2)
+            Mult -> Right (val1 * val2)
+            Div  -> 
+               if val2 == 0 
+               then Left ZeroDiv
+               else Right (val1 / val2)
+            Pow -> Right (val1 ** val2)
```

Comment:
Notice that this case can function the same way as `Sqrt`, which means it can lead to the same errors